drv/lpc55-update-server/src/main.rs

```diff
@@ -279,31 +277,133 @@ fn validate_header_block(
+        while !flash.poll_write_result() {
+            // spin - supposed to be very quick in hardware.
+        }
+    }
+
+    // Now, program the whole page into non-volatile storage. This again uses
+    // page indices, requiring a divide-by-16.
+    flash.start_program(write_addr / WORDSZ);
+    wait_for_erase_or_program(&mut flash);
+
+    Ok(())
+}
+
+/// Utility function that does an interrupt-driven poll and sleep while the
+/// flash controller finishes a write or erase.
+fn wait_for_erase_or_program(flash: &mut drv_lpc55_flash::Flash) -> Result<(), UpdateError> {
+    loop {
+        if let Some(result) = flash.poll_erase_or_program_result() {
+            return result.map_err(|_| UpdateError::FlashError);
+        }
+
+        flash.enable_interrupt_sources();
+        sys_irq_control(notifications::FLASH_IRQ_MASK, true);
+        // RECV from the kernel cannot produce an error, so ignore it.
+        let _ = sys_recv_closed(&mut [], notifications::FLASH_IRQ_MASK, TaskId::KERNEL);
+        flash.disable_interrupt_sources();
+    }
+}
+
+fn same_image(which: UpdateTarget) -> bool {
+    get_base(which) == unsafe { __this_image.as_ptr() as u32 }
+}
+
+/// Returns the byte address of the first byte of the given flash target slot,
+/// or panics if you're holding it wrong.
+fn get_base(which: UpdateTarget) -> u32 {
+    match which {
+        UpdateTarget::ImageA => unsafe { __IMAGE_A_BASE.as_ptr() as u32 },
+        UpdateTarget::ImageB => unsafe { __IMAGE_B_BASE.as_ptr() as u32 },
+        UpdateTarget::Bootloader => unsafe { __IMAGE_STAGE0_BASE.as_ptr() as u32 },
+        _ => unreachable!(),
     }
 }
 
+fn get_end(which: UpdateTarget) -> u32 {
+    match which {
+        UpdateTarget::ImageA => unsafe { __IMAGE_A_END.as_ptr() as u32 },
+        UpdateTarget::ImageB => unsafe { __IMAGE_B_END.as_ptr() as u32 },
+        UpdateTarget::Bootloader => unsafe { __IMAGE_STAGE0_END.as_ptr() as u32 },
+        _ => unreachable!(),
+    }
+}
+
+/// Computes the byte address of the first byte in a particular (slot, page)
+/// combination.
+///
+/// `image_target` designates the flash slot and must be `ImageA`, `ImageB`, or
+/// `Bootloader`, despite containing many other variants. All other choices will
+/// panic. (TODO: fix this when time permits.)
+///
+/// `page_num` designates a flash page (called a block elsewhere in this file, a
+/// 512B unit) within the flash slot. If the page is out range for the target
+/// slot, returns `None`.
+fn target_addr(
+    image_target: UpdateTarget,
+    page_num: u32,
+) -> Option<u32> {
+    let base = get_base(image_target);
+
+    // This is safely calculating addr = base + page_num * PAGE_SIZE
+    let addr = page_num
+        .checked_mul(BLOCK_SIZE_BYTES as u32)?
+        .checked_add(base)?;
+
+    // check addr + PAGE_SIZE < end
+    if addr
+        .checked_add(BLOCK_SIZE_BYTES as u32)?
+        > get_end(image_target)
+    {
+        return None;
+    }
```

Comment:
The mismatch between the comment and code here is confusing, and I suspect the code is wrong.

`get_end` gives the end of an **exclusive** range, so I think we want

```suggestion
    // check addr + PAGE_SIZE < end
    if addr
        .checked_add(BLOCK_SIZE_BYTES as u32)?
        >= get_end(image_target)
    {
        return None;
    }
```

Reply:
I spent some time staring at this after I simplified the code (the original, if you're curious, is `lib/lpc55-flash/src/lib.rs:119`). I believe it is the _comment_ that is wrong. If changed to `>=`, this would prevent you from talking about the last page.

I will fix the comment.